src/Microsoft.AspNet.Mvc.WebApiCompatShim/Conventions/WebApiParameterConventionsApplicationModelConvention.cs

```diff
@@ -15,30 +16,36 @@ public void Apply(ActionModel action)
         {
             if (IsConventionApplicable(action.Controller))
             {
+                var optionalParameters = new HashSet<string>();
                 foreach (var parameter in action.Parameters)
                 {
-                    if (parameter.BinderMetadata is IBinderMetadata)
+                    // Some IBindingSourceMetadata attributes like ModelBinder attribute return null 
+                    // as their binding source. Special case to ensure we do not ignore them.
+                    if (parameter.BindingInfo.BindingSource != null || 
+                        parameter.Attributes.OfType<IBindingSourceMetadata>().Any())
                     {
                         // This has a binding behavior configured, just leave it alone.
                     }
                     else if (ValueProviderResult.CanConvertFromString(parameter.ParameterInfo.ParameterType))
                     {
                         // Simple types are by-default from the URI.
-                        parameter.BinderMetadata = new FromUriAttribute();
+                        parameter.BindingInfo.BindingSource = (new FromUriAttribute()).BindingSource;
                     }
                     else
                     {
                         // Complex types are by-default from the body.
-                        parameter.BinderMetadata = new FromBodyAttribute();
+                        parameter.BindingInfo.BindingSource = (new FromBodyAttribute()).BindingSource;
                     }
 
                     // If the parameter has a default value, we want to consider it as optional parameter by default.
-                    var optionalMetadata = parameter.BinderMetadata as FromUriAttribute;
-                    if (parameter.ParameterInfo.HasDefaultValue && optionalMetadata != null)
+                    var optionalMetadata = parameter.Attributes.OfType<IOptionalBinderMetadata>().SingleOrDefault();
```

Comment:
You need to check the value of `optionalMetadata.IsOptional`. There should at least be a failing test for this.